Capsule protectli

wiktormowinski opened the pull request:
coreboot part of https://github.com/Dasharo/dasharo-issues/issues/1390, https://github.com/Dasharo/edk2/pull/254 is for EDK.

*ref: prot-1829*

wiktormowinski committed:
payloads/edk2: add an option for graphical capsule update

Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Wiktor Mowinski <wiktor.mowinski@3mdeb.com>

wiktormowinski committed:
configs/config.protectli_vp66xx: enable capsule update on Protectli ADL

Change-Id: Ie4a013639177dc209c9592ad0f5b72c134252a36
Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Wiktor Mowinski <wiktor.mowinski@3mdeb.com>

wiktormowinski committed:
.github/workflows/build.yml: add capsule update firmware build runner action

Upstream-Status: Inappropriate [Dasharo downstream]
Signed-off-by: Wiktor Mowinski <wiktor.mowinski@3mdeb.com>

SergiiDmytruk commented on configs/config.protectli_vp66xx:
```diff
@@ -22,6 +22,11 @@ CONFIG_DASHARO_PREFER_S3_SLEEP=y
+CONFIG_DRIVERS_EFI_FW_INFO=y
+CONFIG_DRIVERS_EFI_MAIN_FW_GUID="dbe538fe-d0fa-4a33-aa36-1d32945e4154"
+CONFIG_DRIVERS_EFI_MAIN_FW_VERSION=0x00090180
+CONFIG_DRIVERS_EFI_MAIN_FW_LSV=0x00090080
```
You want to bump `CONFIG_LOCALVERSION` to `v0.9.2-rc0` and set these two options to `0x00090200`.  RC is set to `80` when a release is being finalized.

wiktormowinski replied:
sure: https://github.com/Dasharo/coreboot/pull/696/commits/7de739c1d599d0102ea2372c22f91a4b12279ac6

SergiiDmytruk replied:
Given that v0.9.2 has just been released (https://docs.dasharo.com/variants/protectli_vp66xx/releases/#v092-2025-06-17), a different values might need to be used.  But need to figure out correct target branch first.

SergiiDmytruk replied:
I think we can use v0.9.3-rc0 and bump to another version later if necessary, so:
```
CONFIG_LOCALVERSION="v0.9.3-rc0"
...
CONFIG_DRIVERS_EFI_MAIN_FW_VERSION=0x00090300
CONFIG_DRIVERS_EFI_MAIN_FW_LSV=0x00090300
```
Previous releases didn't have capsules, so LSV and VERSION should be equal.

wiktormowinski replied:
alright https://github.com/Dasharo/coreboot/pull/696/commits/5dea72afd0256f44605b5322c1f40540b0d7074a

SergiiDmytruk commented on .github/workflows/build.yml:
```diff
@@ -159,12 +159,18 @@ jobs:
       - name: Build Dasharo
         run: |
           ./build.sh ${{ matrix.model }}
+          pkcs7sign=payloads/external/edk2/workspace/Dasharo/BaseTools/Source/Python/Pkcs7SignAdd commentMore actions
```
Some typo:
```suggestion
          pkcs7sign=payloads/external/edk2/workspace/Dasharo/BaseTools/Source/Python/Pkcs7Sign
```

wiktormowinski replied:
here: https://github.com/Dasharo/coreboot/pull/696/commits/de79468e764954a7717ea36b66004e6d1aed23e9

SergiiDmytruk commented on payloads/external/Makefile.mk:
```diff
@@ -290,7 +290,8 @@ $(obj)/UEFIPAYLOAD.fd: $(DOTCONFIG) $(IPXE_EFI)
 		CONFIG_DGPU_ONLY_AVAILABLE=$(CONFIG_DGPU_ONLY_AVAILABLE) \
 		CONFIG_EDK2_FTDI_USB_UART_SUPPORT=$(CONFIG_EDK2_FTDI_USB_UART_SUPPORT) \
-		CONFIG_QEMU_PLATFORM=$(CONFIG_CPU_QEMU_X86)
+		CONFIG_QEMU_PLATFORM=$(CONFIG_CPU_QEMU_X86)\
```
nit:
```suggestion
		CONFIG_QEMU_PLATFORM=$(CONFIG_CPU_QEMU_X86) \
```
It's unusual to not have a space there.

wiktormowinski replied:
it seemed a bit off, now i know what was wrong: https://github.com/Dasharo/coreboot/pull/696/commits/44852e97f1f4af7d389f9b170398c6e6680fdcbd

SergiiDmytruk commented on payloads/external/edk2/Kconfig.dasharo:
```diff
@@ -392,4 +392,11 @@ config EDK2_ENABLE_QUIET_BOOT_FEATURE
 	help
 	  Enables quiet boot option in the TianoCore Payload.
 
+config EDK2_GRAPHICAL_UPDATE_DISPLAY
```
Maybe name it like this to make it easier to understand out of context (better to have consistent naming in EDK and coreboot):
```suggestion
config EDK2_GRAPHICAL_CAPSULE_PROGRESS
```

wiktormowinski replied:
thanks, i can never get the names right: https://github.com/Dasharo/coreboot/pull/696/commits/9242f1a7e05d48c442436fd43c92553eba413932

SergiiDmytruk commented on payloads/external/edk2/Kconfig.dasharo:
```diff
@@ -392,4 +392,11 @@ config EDK2_ENABLE_QUIET_BOOT_FEATURE
 
+config EDK2_GRAPHICAL_UPDATE_DISPLAY
+	bool "Enable display of capsule update progress bar via graphical interface"
+	default n
```
Having it off by default changes behaviour of capsules for MSI and NovaCustom.  This either needs to be on by default and disabled for VP66XX or need to turn it on in 13 configs which enable capsules.  Don't know which state will prove to be more popular in the future, but so far it's easier to have this on by default.

wiktormowinski replied:
makes sense, added config line in protectli config: https://github.com/Dasharo/coreboot/pull/696/files#diff-9a72f6ce694cf3d77e66993ade0a81b80999f603c47c63e19213c5e05bb3bf72 
and changed logic in edk2 PR: https://github.com/Dasharo/edk2/pull/254/commits/3d36cf5cb9a7a488abb51defa7bc6e4d805ebf1f

SergiiDmytruk replied:
I meant using `default y` and doing `CONFIG_EDK2_GRAPHICAL_CAPSULE_PROGRESS=y` in the config.  No need to change help text or swap lines in EDK.

SergiiDmytruk replied:
> I meant using `default y` and doing `CONFIG_EDK2_GRAPHICAL_CAPSULE_PROGRESS=y` in the config.

A typo, Protectli config should have `CONFIG_EDK2_GRAPHICAL_CAPSULE_PROGRESS=n`.

wiktormowinski replied:
at first i thought the same, but then the CI had some objections, so i think something else should be done to make it work: https://github.com/Dasharo/coreboot/actions/runs/15826924259/job/44609431807?pr=696

SergiiDmytruk replied:
CI is right.  I guess defconfigs prefer `# CONFIG_* is not set` rather than `n` for some reason, the effect is the same.

SergiiDmytruk reviewed (changes requested):
I've missed it earlier, but config also needs to point at EDK's commit from https://github.com/Dasharo/edk2/pull/254.

wiktormowinski commented:
> I've missed it earlier, but config also needs to point at EDK's commit from https://github.com/Dasharo/edk2/pull/254.

@SergiiDmytruk i did it locally during build but i can push it too: https://github.com/Dasharo/coreboot/pull/696/commits/77c9b78441d892ddd53ca6b3ac577e1f2ff3c5ba

SergiiDmytruk commented on .github/workflows/build.yml:
```diff
@@ -159,12 +159,18 @@ jobs:
       - name: Build Dasharo
         run: |
           ./build.sh ${{ matrix.model }}
+          pkcs7sign=payloads/external/edk2/workspace/Dasharo/BaseTools/Source/Python/Pkcs7Sign
+          ./capsule.sh make -t "$pkcs7sign"/TestRoot.pub.pem \
+                            -o "$pkcs7sign"/TestSub.pub.pem \
+                            -s "$pkcs7sign"/TestCert.pem
+          mv *.cap build/coreboot.cap
```
CI fails because this block of new lines needs to be run conditionally:
```bash
if grep -q CONFIG_DRIVERS_EFI_UPDATE_CAPSULES=y .config; then
    # ...
fi
```
It's different from MSI because only one of the configs has capsules enabled.

wiktormowinski replied:
sure: https://github.com/Dasharo/coreboot/pull/696/commits/8179d303050bbf8707558d95683f727b826f0714
now CI works

SergiiDmytruk replied:
nit: missing indentation level for the contents between `if` and `fi`.

wiktormowinski replied:
https://github.com/Dasharo/coreboot/pull/696/commits/5c76dc1258deb2a6b82e66fa2d4f3bb8de9ed99a

SergiiDmytruk commented:
> > I've missed it earlier, but config also needs to point at EDK's commit from [Dasharo/edk2#254](https://github.com/Dasharo/edk2/pull/254).
> 
> @SergiiDmytruk i did it locally during build but i can push it too: [77c9b78](https://github.com/Dasharo/coreboot/commit/77c9b78441d892ddd53ca6b3ac577e1f2ff3c5ba)

It's annoying to have hashes synchronized between PRs, but that also allows using binaries built by the CI and makes CI test both parts combined.

SergiiDmytruk reviewed (changes requested):
Some final notes.

SergiiDmytruk reviewed (dismissed):
I was checking handling of the option locally and cleaned up history while making a couple corrections (some things needed to be inverted after changing meaning of options).

If testing on hardware produces expected results, can merge this once CI is green.

P.S. All commits are `Inappropriate [Dasharo downstream]` because configs, CI and new EDK option are specific to Dasharo.

SergiiDmytruk committed:
intel/fast_spi/mmap_boot.c: don't add empty window

When flash chip fully fits into the main mmap window, don't create an
empty window backed up by RW device.

Upstream-Status: Pending
Change-Id: I925c1f9a1024e702ba0048037d4ea2f822488be1
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
intel/fast_spi/fast_spi.c: extend BIOS to fill fixed window

Platforms with 16 MiB chips don't start BIOS at offset 0, making it
possible to fit entire chip into the fixed memory mapped window.

Upstream-Status: Pending
Change-Id: Iddb819f20906eaa392caf2cfc63aebf8354e2049
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
drivers/efi: require FUM for capsules

Capsules were implemented with the assumption that FUM is available,
enforce the assumption.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I9a1bc73a425f66aed5490f2c5bf3d067dfb505a1
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
payloads/edk2/Kconfig.dasharo: bump for GRAPHICAL_CAPSULE_PROGRESS

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: Idb4d2ed1a715dc01697d12185838afd4c41186b4
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>
